### PR TITLE
[TRTLLM-5966][feat] Helix: add alltoall op

### PR DESCRIPTION
**Pull request by MatthiasKohl**
## Summary by CodeRabbit

* **New Features**
  * Added a distributed all-to-all operation to tensorrt_llm._torch.distributed with a CUDA backend and CPU/no-op fallback; exposed as a PyTorch-compatible alltoall and included in the public API.
  * Supports single or multiple tensors with configurable split dimensions and optional stacking/concatenation; build now includes the all-to-all implementation.

* **Tests**
  * Added multi-GPU integration tests (2‑GPU and 4‑GPU) validating correctness across shapes, dtypes, and partition/stacking configurations.

## Description

Adds an all-to-all op required for Helix parallelism support

## Test Coverage

tests/unittest/_torch/multi_gpu/test_alltoall.py

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tests/unittest/_torch/multi_gpu/test_alltoall.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    for rank in range(world_size):
+        input_tensors.append([])
+        expected_send_tensors.append([])
+        for list_idx in range(num_lists):
+            # Each rank creates a tensor with unique data
+            tensor = torch.randn(*shape, dtype=dtypes[list_idx])
+            input_tensors[-1].append(tensor)
+            d = dims[list_idx] if isinstance(dims, list) else dims
+            send_tensors = []
+            for r in range(world_size):
```
Comment (brb-nv) [inline tests/unittest/_torch/multi_gpu/test_alltoall.py:108]: I'm a bit confused about what's going on here. Can you please leave a comment explaining what's happening in this `for` loop?

File: tensorrt_llm/_torch/distributed/ops.py [hunk 2]
```diff
@@ -225,6 +225,84 @@ def convert_output(x, x_info):
...
+    for inp, dim in zip(inputs, dims):
+        size_per_rank, rem = divmod(inp.shape[dim], n_ranks)
+        assert rem == 0, \
+            f"input.shape[{dim}] must be divisible by n_ranks ({n_ranks}), but got shape {inp.shape}"
+
+        # we split the input into n_ranks equal parts along dimension `dim`
+        # note: the requirement for the C++ op is that all parts have same shape,
+        # dtype and device. If we make sure that size at `dim` evenly divides into
+        # `n_ranks`, all the conditions are met
+        op_inputs.extend(torch.split(inp, size_per_rank, dim=dim))
```
Comment (brb-nv) [inline tensorrt_llm/_torch/distributed/ops.py:281]: Noob question: Can you please explain when this split would be needed?
Each KVP GPU has its own local softmax statistics, right (say max of local exps and sum of local exps)?
Why would be want to split each of those stats further into n_ranks parts? Maybe I'm missing something.

**Review by brb-nv — APPROVED** [review]
Comments are mostly questions I'm asking for my own understanding.

File: cpp/tensorrt_llm/thop/alltoallOp.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        std::vector<torch::Tensor> output_list;
+        for (int il = 0; il < num_lists_; ++il)
+        {
+            auto off = il * num_ranks;
+            auto type = tensorrt_llm::runtime::TorchUtils::dataType(input_list[off].scalar_type());
+            auto nccl_type = (*getDtypeMap())[type];
+            for (int r = 0; r < num_ranks; ++r)
+            {
+                auto const& input = input_list[off + r];
+                auto output = torch::empty_like(input);
```
[the remark this answers is not in the document]
Comment (MatthiasKohl) [reply cpp/tensorrt_llm/thop/alltoallOp.cpp:78]: it doesn't have to be the case for NCCL comms, it depends on the exact operation.
But for a send-receive like here, it basically must be the case because the receiving ranks must receive the same amount of data that the sending ranks sent. In `alltoall`, both the sending and receiving ranks are always all ranks, so they are the same. Because of this, all tensors will be the same on the receiving end as on the sending end for all ranks.

File: tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py [hunk 4]
```diff
@@ -531,3 +531,16 @@ def _(router_logits, topk):
         return router_logits.new_empty(
             sz, dtype=torch.int32), router_logits.new_empty(sz,
                                                             dtype=torch.float32)
+
+    @torch.library.register_fake("trtllm::alltoall")
```
[the remark this answers is not in the document]
Comment (MatthiasKohl) [reply tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py:535]: IIUC, it's mainly for compiler tracing like `torch.compile` where it avoids graph breaks

File: tests/unittest/_torch/multi_gpu/test_alltoall.py (hunk 1 above)
Comment (MatthiasKohl) [reply tests/unittest/_torch/multi_gpu/test_alltoall.py:114]: done, we split up the tensors to calculate the expected output from alltoall

File: tensorrt_llm/_torch/distributed/ops.py (hunk 2 above)
Comment (MatthiasKohl) [reply tensorrt_llm/_torch/distributed/ops.py:281]: the reason for the alltoall in Helix is to re-shard the data for the output projection, where the KVP ranks are now TP ranks. This is done across the head dimension.
So we split the head dimension across the KVP ranks and re-distribute the individual parts, s.t. they now own a sub-part of the heads, as expected by the output projection where we now have OUT_TP := IN_TP * KVP and thus the head dimension must be split into IN_TP * KVP parts

File: tensorrt_llm/_torch/distributed/ops.py [hunk 5]
```diff
@@ -225,6 +225,84 @@ def convert_output(x, x_info):
     return output
 
 
+def alltoall(
```
Comment (bobboli) [inline tensorrt_llm/_torch/distributed/ops.py:228]: Maybe rename this OP to `alltoall_helix` since it has assumptions that narrow the general alltoall inferface: the comm pattern is fixed by uniformly slicing at send rank and assembling at recv rank, given the dimensions.

File: tensorrt_llm/_torch/distributed/ops.py [hunk 6]
```diff
@@ -225,6 +225,84 @@ def convert_output(x, x_info):
...
+        # we split the input into n_ranks equal parts along dimension `dim`
+        # note: the requirement for the C++ op is that all parts have same shape,
+        # dtype and device. If we make sure that size at `dim` evenly divides into
+        # `n_ranks`, all the conditions are met
+        op_inputs.extend(torch.split(inp, size_per_rank, dim=dim))
+
+    # note: we need to ensure that the input tensors are contiguous to provide
+    # the right data pointers to the C++ op
+    op_inputs = [op_inp.contiguous() for op_inp in op_inputs]
+    outputs = torch.ops.trtllm.alltoall(
```
Comment (bobboli) [inline tensorrt_llm/_torch/distributed/ops.py:286]: Same as above.

**Review by bobboli — APPROVED** [review]

File: tensorrt_llm/_torch/distributed/ops.py [hunk 7]
```diff
@@ -225,6 +225,84 @@ def convert_output(x, x_info):
...
+        op_inputs,
+        group,
+        num_lists=len(inputs),
+    )
+
+    outputs_split = [
+        outputs[i * n_ranks:(i + 1) * n_ranks] for i in range(len(inputs))
+    ]
+    for i in range(len(inputs)):
```
Comment (bobboli) [inline tensorrt_llm/_torch/distributed/ops.py:295]: The final concat/stack might be costly. Is it possible to pre-allocate complete recv tensors in advance, and pass the offsetted pointers to nccl? Might not work with concat though since the invidivual sub tensors are not contiguous.

**Review by bobboli — APPROVED** [review]

File: tensorrt_llm/_torch/distributed/ops.py (hunk 2 above)
Comment (bobboli) [reply tensorrt_llm/_torch/distributed/ops.py:281]: BTW, what is KVP? KV partition?

File: tensorrt_llm/_torch/distributed/ops.py (hunk 7 above)
Comment (MatthiasKohl) [reply tensorrt_llm/_torch/distributed/ops.py:295]: Yes this is possible, but only for specific dimensions.

IMO, performance optimizations should be done in a follow-up PR, this PR is not about performance but functionality.

Comment (MatthiasKohl) [reply tensorrt_llm/_torch/distributed/ops.py]: I updated the op to be more generic, and also included this change now, so this should be resolved

**Comment by brb-nv** [comment]
Previous unrelated failure waived an hour ago.
https://github.com/NVIDIA/TensorRT-LLM/pull/7812

File: tensorrt_llm/_torch/distributed/ops.py [hunk 8]
```diff
@@ -229,6 +229,49 @@ def convert_output(x, x_info):
...
+    num_lists = len(inputs) // n_ranks
+    outputs = []
+    for il in range(num_lists):
+        ref_input = inputs[il * n_ranks]
+        assert all([inputs[i].shape == ref_input.shape for i in range(il * n_ranks + 1, (il + 1) * n_ranks)]),\
+            "all input tensors in a group must have the same shape"
+        outputs.append(ref_input.new_empty((n_ranks, ) + ref_input.shape))
+
+    output_list = [out[i] for out in outputs for i in range(n_ranks)]
+    torch.ops.trtllm.alltoall(output_list, inputs, group, num_lists)
```
Comment (liji-nv) [inline tensorrt_llm/_torch/distributed/ops.py:271]: I notice that you are using slicing. If you are handling everything in the C++, I think the op can directly return the expected outputs.

File: cpp/tensorrt_llm/thop/alltoallOp.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    op.initialize();
+    op.run(output_list, input_list, num_lists);
+#endif // ENABLE_MULTI_DEVICE
+}
+
+} // namespace torch_ext
+
+TORCH_LIBRARY_FRAGMENT(trtllm, m)
+{
+    m.def("alltoall(Tensor[] output_list, Tensor[] input_list, int[] group, int? num_lists) -> ()");
```
[the remark this answers is not in the document]
Comment (MatthiasKohl) [reply cpp/tensorrt_llm/thop/alltoallOp.cpp:115]: The advantage of have the output_list as a mutable parameter is that we can pass any kind of view or tensor from python: it is more flexible on the python side. If the C++ method allocates, then we will have an explicit assumption of the memory layout from within C++, which is harder to modify.
Does this make sense?

[the remark this answers is not in the document]
Comment (liji-nv) [reply cpp/tensorrt_llm/thop/alltoallOp.cpp:115]: The problem is list of mutate tensor cannot be correctly handled by our torch.compile workflow as of today.

**Review by bobboli — APPROVED** [review]

**Review by liji-nv — APPROVED** [review]
